File: Enhancedwindow.py
import cv2
import numpy as np

class EnhancedWindow:
    def __init__(self, x, y, width, height, title, color=(60, 60, 60), title_color=(100, 100, 255)):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.title = title
        self.color = color
        self.title_color = title_color
        self.minimized = False
        self.dragging = False
        self.offset_x = 0
        self.offset_y = 0
        self.title_bar_height = 35

    def inside(self, mx, my):
        """Check if mouse is inside the full window area"""
        return self.x <= mx <= self.x + self.width and self.y <= my <= self.y + self.height

    def handle_mouse(self, event, mx, my, flags):
        """Handle drag and minimize button"""
        # Minimize button area (top-right corner)
        minimize_btn_area = (self.x + self.width - 30, self.y, self.x + self.width, self.y + self.title_bar_height)

        if event == cv2.EVENT_LBUTTONDOWN:
            if minimize_btn_area[0] <= mx <= minimize_btn_area[2] and minimize_btn_area[1] <= my <= minimize_btn_area[3]:
                self.minimized = not self.minimized
            elif self.inside(mx, my):
                self.dragging = True
                self.offset_x = mx - self.x
                self.offset_y = my - self.y

        elif event == cv2.EVENT_LBUTTONUP:
            self.dragging = False

        elif event == cv2.EVENT_MOUSEMOVE and self.dragging:
            self.x = mx - self.offset_x
            self.y = my - self.offset_y

    def draw(self, frame):
        """Draw the window"""
        h, w = frame.shape[:2]
        self.x = np.clip(self.x, 0, w - self.width)
        self.y = np.clip(self.y, 0, h - self.title_bar_height)

        # Draw title bar
        cv2.rectangle(frame, (self.x, self.y), (self.x + self.width, self.y + self.title_bar_height), self.title_color, -1)

        # Draw minimize/expand button
        cv2.rectangle(frame, (self.x + self.width - 30, self.y),
                      (self.x + self.width, self.y + self.title_bar_height), (80, 80, 200), -1)
        # Draw horizontal line instead of "-" character
        btn_symbol = '+' if self.minimized else '-'
        cv2.putText(frame, btn_symbol, (self.x + self.width - 20, self.y + 25),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)
        
        # Draw title text (optional, can be removed)
        cv2.putText(frame, self.title, (self.x + 10, self.y + 25),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)

        # Draw content area if not minimized
        if not self.minimized:
            cv2.rectangle(frame, (self.x, self.y + self.title_bar_height),
                          (self.x + self.width, self.y + self.height), self.color, -1)


def main():
    window_name = "Enhanced OpenCV UI"
    cv2.namedWindow(window_name)
    frame = np.zeros((600, 1000, 3), dtype=np.uint8)
    frame[:] = (35, 35, 35)

    panel = EnhancedWindow(100, 100, 400, 300, "Panel", color=(50, 50, 70), title_color=(90, 130, 255))
    options = EnhancedWindow(550, 150, 400, 300, "Options", color=(70, 50, 50), title_color=(255, 90, 90))

    def mouse_callback(event, x, y, flags, param):
        panel.handle_mouse(event, x, y, flags)
        options.handle_mouse(event, x, y, flags)

    cv2.setMouseCallback(window_name, mouse_callback)

    while True:
        frame[:] = (35, 35, 35)
        panel.draw(frame)
        options.draw(frame)

        cv2.imshow(window_name, frame)
        key = cv2.waitKey(20)
        if key == 27:  # ESC
            break

    cv2.destroyAllWindows()


if __name__ == "__main__":
    main()

File: Enhancedwindow.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

class EnhancedWindow {
public:
    int x, y, width, height;
    string title;
    Scalar color;
    Scalar title_color;
    bool minimized;
    bool dragging;
    int offset_x, offset_y;
    int title_bar_height;

    EnhancedWindow(int _x, int _y, int _width, int _height, string _title,
                   Scalar _color = Scalar(60, 60, 60),
                   Scalar _title_color = Scalar(100, 100, 255))
    {
        x = _x;
        y = _y;
        width = _width;
        height = _height;
        title = _title;
        color = _color;
        title_color = _title_color;
        minimized = false;
        dragging = false;
        offset_x = 0;
        offset_y = 0;
        title_bar_height = 35;
    }

    bool inside(int mx, int my) {
        return (mx >= x && mx <= x + width && my >= y && my <= y + height);
    }

    void handleMouse(int event, int mx, int my, int flags) {
        Rect minimize_btn_area(x + width - 30, y, 30, title_bar_height);

        if (event == EVENT_LBUTTONDOWN) {
            if (minimize_btn_area.contains(Point(mx, my))) {
                minimized = !minimized;
            }
            else if (inside(mx, my)) {
                dragging = true;
                offset_x = mx - x;
                offset_y = my - y;
            }
        }
        else if (event == EVENT_LBUTTONUP) {
            dragging = false;
        }
        else if (event == EVENT_MOUSEMOVE && dragging) {
            x = mx - offset_x;
            y = my - offset_y;
        }
    }

    void draw(Mat& frame) {
        int frame_w = frame.cols;
        int frame_h = frame.rows;

        x = max(0, min(x, frame_w - width));
        y = max(0, min(y, frame_h - title_bar_height));

        // Title bar
        rectangle(frame, Point(x, y), Point(x + width, y + title_bar_height), title_color, FILLED);

        // Minimize/Expand button
        rectangle(frame, Point(x + width - 30, y),
                  Point(x + width, y + title_bar_height), Scalar(80, 80, 200), FILLED);

        string btn_symbol = minimized ? "+" : "-";
        putText(frame, btn_symbol, Point(x + width - 20, y + 25),
                FONT_HERSHEY_SIMPLEX, 0.7, Scalar(255, 255, 255), 2);

        // Title text
        putText(frame, title, Point(x + 10, y + 25),
                FONT_HERSHEY_SIMPLEX, 0.7, Scalar(255, 255, 255), 2);

        // Content area if not minimized
        if (!minimized) {
            rectangle(frame, Point(x, y + title_bar_height),
                      Point(x + width, y + height), color, FILLED);
        }
    }
};

EnhancedWindow panel(100, 100, 400, 300, "Panel", Scalar(50, 50, 70), Scalar(90, 130, 255));
EnhancedWindow options(550, 150, 400, 300, "Options", Scalar(70, 50, 50), Scalar(255, 90, 90));

void mouseCallback(int event, int x, int y, int flags, void* userdata) {
    panel.handleMouse(event, x, y, flags);
    options.handleMouse(event, x, y, flags);
}

int main() {
    string window_name = "Enhanced OpenCV UI";
    namedWindow(window_name);
    setMouseCallback(window_name, mouseCallback);

    Mat frame(600, 1000, CV_8UC3, Scalar(35, 35, 35));

    while (true) {
        frame.setTo(Scalar(35, 35, 35));

        panel.draw(frame);
        options.draw(frame);

        imshow(window_name, frame);
        char key = (char)waitKey(20);
        // Press ESC to exit
        if (key == 27) break; 
    }

    destroyAllWindows();
    return 0;
}
